### PR TITLE
Document amoc coordinator

### PR DESCRIPTION
**Pull request by aleklisi**
This PR attempts to document `amoc_coordinator` module.

### CODE REVIEW

File: doc/amoc_coordinator.md [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+
+## Module
+
+`amoc_coordinator` 
+
+## Description
+
+This module allows to synchronize the users, and act on groups of them.
+
+The coordinator reacts to new users showing up in a system, according to `Coordination Plan`. `Coordination Plan` consists of `Coordination Items`, each of them is defined as one of the following: `{NumberOfUsers, CoordinationActions}`. When `Number Of Users` is set to `all`, then only `Coordination Actions` with the arities `/1, /2` are handled. When `Number Of Users` is set to a positive integer, then when then all `Coordination Actions` with the arities `/1, /2` and `/3` are handled. When new batch size `NumberOfUsers` of users calls `add` to the coordinator, then `Coordination Actions` following the `NumberOfUsers` in the `Coordination Plan` are triggered. When there is more then one `Coordination Items` matching the `NumberOfUsers` all are triggered, Each will be passed respective number of users, eg if the `Coordination Plan` is `[{2, Act1}, {3, Act2}]` then on 6th users calling `add`, `Act1` will be called with 2 users passed and `Act2` wil be called with 3 users passed.
```
[the remark this answers is not in the document]
Comment (michalwski) [reply doc/amoc_coordinator.md]: Also this paragraph is a bit hard to read, more things were uncovered for me when reading the example. Do you think this paragraph could be reworked?

File: doc/amoc_coordinator.md [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+User :: {UsersPid :: pid(), Data :: any()}
+Timeout :: pos_integer() | infinity
+```
+
+This function starts a coordinator. Usually is called in `init/0` of a amoc scenario.
+
+### `add(CoordinatorName, Data) -> ok`
+### `add(CoordinatorName, Pid, Data) -> ok`
+
+### Types
```
[the remark this answers is not in the document]
Comment (aleklisi) [reply doc/amoc_coordinator.md]: Yes

File: doc/amoc_coordinator.md [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+
+ok  % result of calling amoc_local:do/3
+
+User = 2 % First user is started
+
+% We have 2 users added to amoc_coordinator so all of actions {2, _} are triggered:
+Two new users showed up: Event = coordinate; User1 = {<0.1142.0>,2}; User2 = {<0.1140.0>,1}
+% This action triggers sending {hello,Id} to the users 1 and 2
+Two new users showed up: Event = coordinate; ListOfUsers = [{<0.1142.0>,2},{<0.1140.0>,1}]
+Two new users showed up: Event = coordinate
```
[the remark this answers is not in the document]
Comment (aleklisi) [reply doc/amoc_coordinator.md]: The order of executions is always reversed with reference to order on the `CoordinationPlan`, I shall add a piece of information about that.

[the remark this answers is not in the document]
Comment (aleklisi) [reply doc/amoc_coordinator.md]: I do not know why is that.

File: doc/amoc_coordinator.md [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+            end},
+        {2, fun(Event, User1, User2) ->
+                io:fwrite("Two new users showed up: Event = ~p; User1 = ~p; User2 = ~p\n", [Event, User1, User2])
+            end},
+        
+        {3, fun(_Event) ->
+                io:fwrite("Three new users showed up\n", [])
+            end},
+        {all, fun(Event) ->
+                io:fwrite("All users have logged in Event = ~p\n", [Event])
```
[the remark this answers is not in the document]
Comment (aleklisi) [reply doc/amoc_coordinator.md]: It means, all users called `amoc_coordinator:add/2` or `amoc_coordinator:add/3`.
I shall replace `logged` with `called amoc_coordinator:add`.

File: doc/amoc_coordinator.md [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+#### `add(CoordinatorName, Pid, Data) -> ok`
+
+##### Types
+```erlang
+CoordinatorName :: atom()
+UsersPid :: pid()
+Data :: any()
+```
+
+This function adds current process data. Usually is called in `start/2` of a amoc scenario.
```
Comment (michalwski) [inline doc/amoc_coordinator.md]: I think it's important to say that this function basically increases number of already started users. Without this the coordination actions may not be triggered, especially the one for all users. If for some reason one session doesn't call the function and we wanted to have 10 users in the test, then the rule for `all` will not be triggered. Am I right?

Comment (aleklisi) [reply doc/amoc_coordinator.md]: No

Comment (aleklisi) [reply doc/amoc_coordinator.md]: If we have 10 users in a test but only 5 of them calls the function `add`. and the event `timeout` happens action for `all` will be triggered for the 5 users who registered to the amoc coordinator.

File: doc/amoc_coordinator.md [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+
+This module allows to synchronize the users, and act on groups of them.
+
+The coordinator reacts to new users showing up in a system, according to the `Coordination Plan`. 
+The `Coordination Plan` consists of `Coordination Items`, and each of them is defined as one of the following: `{NumberOfUsers, CoordinationActions}`.
+ - When the `Number Of Users` is set to `all`, then only `Coordination Actions` with the arities `/1, /2` are handled. 
+ The `Coordination Items` with `all` are triggered by the `timeout` event. 
+ - When the `Number Of Users` is set to a positive integer, all `Coordination Actions` with arities `/1, /2` and `/3` are handled.
+
+The timeout timer is reset by calling function `add`.
```
Comment (goddammit) [inline doc/amoc_coordinator.md]:
```suggestion
The timeout timer is reset by calling the `add` function.
```

File: doc/amoc_coordinator.md [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+If more then one of the `Coordination Items` matching the `NumberOfUsers` is triggered, each of them will be passed a respective number of users.
+For example if the `Coordination Plan` is `[{2, Act1}, {3, Act2}]` then on the 6th user calling `add`, `Act1` will be called with 2 users passed and `Act2` wil be called with 3 users passed.
+
+`Coordination Actions` may be one of the following:
+ - `fun(Event) -> any()` - this type of action does not care about particular users, but only about the number of them;
+ - `fun(Event, ListOfUsersData) -> any()` - this type of action gets `ListOfUsersData` which is a list of tuples `{Pid, Data}` with `Pid`s passed by users calling `amoc_coordinator:add/2` or `amoc_coordinator:add/3`;
+- `fun(Event, User1, User2) -> any()` - this type of action gets `distinct pairs` from the batch of users `User1` and `User2` which are tuples `{Pid, Data}` with `Pid`s passed by users calling `amoc_coordinator:add/2` or `amoc_coordinator:add/3`;
+
+`Coordination Actions` are executed in reversed order in relation to thair declaration in the `Coordination Plan`.
```
Comment (goddammit) [inline doc/amoc_coordinator.md]:
```suggestion
`Coordination Actions` are executed in a reversed order in relation to their declaration in the `Coordination Plan`.
```

File: doc/amoc_coordinator.md [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+##### Types
+```erlang
+CoordinatorName :: atom()
+CoordinationPlan  :: [ CoordinationItem ]
+CoordinationItem :: {NumberOfUsers :: pos_integer() | all, CoordinationActions}
+CoordinationActions :: [ CoordinationActions ] | CoordinationAction
+CoordinationAction :: 
+    fun(Event) -> any() |
+    fun(Event, ListOfUsersData) -> any() |
+    fun(Event, MaybeUser, MaybeUser) -> any()
```
Comment (NelsonVides) [inline doc/amoc_coordinator.md]: I understand that the two users given to this function are ensured to be different, right?
Then maybe you can write
```erlang
    fun(Event, MaybeUser1, MaybeUser2) when MaybeUser1 =/= MaybeUser2 -> any()
```

File: doc/amoc_coordinator.md [hunk 9]
```diff
@@ -0,0 +1,170 @@
+# amoc_coordinator
+
+## Module
+
+`amoc_coordinator` 
+
+## Description
+
+This module allows to synchronize the users, and act on groups of them.
```
Comment (erszcz) [inline doc/amoc_coordinator.md]:
```suggestion
This module allows to synchronize the users and act on groups of them.
```

File: doc/amoc_coordinator.md [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+
+`Coordination Actions` are executed in a reversed order in relation to their declaration in the `Coordination Plan`.
+
+`distinct pairs` - in the context, these are pairs from collection where:
+ - when `{A, B}` is in the `distinct pairs` then `{B, A}` is not;
+ - when `{A, A}` is not in the `distinct pairs`;
+ - all pairs are distinct;
+ - Eg. for `[a]` `distinct pairs` collection is `[{a, undefined}]`;
+ - Eg. for `[a, b]` `distinct pairs` collection is `[{a, b}]`;
+ - Eg. for `[a, b, c]` `distinct pairs` collection is `[{a, b}, {a, c}, {b, c}]`.
```
Comment (NelsonVides) [inline doc/amoc_coordinator.md]: > – Eg. for `xxx`, the `distinct pairs` collection is `yyy`

(comma and article missing)

File: doc/amoc_coordinator.md [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+If more then one of the `Coordination Items` matching the `NumberOfUsers` is triggered, each of them will be passed a respective number of users.
+For example if the `Coordination Plan` is `[{2, Act1}, {3, Act2}]` then on the 6th user calling `add`, `Act1` will be called with 2 users passed and `Act2` wil be called with 3 users passed.
+
+`Coordination Actions` may be one of the following:
+ - `fun(Event) -> any()` - this type of action does not care about particular users, but only about the number of them;
+ - `fun(Event, ListOfUsersData) -> any()` - this type of action gets `ListOfUsersData` which is a list of tuples `{Pid, Data}` with `Pid`s passed by users calling `amoc_coordinator:add/2` or `amoc_coordinator:add/3`;
+- `fun(Event, User1, User2) -> any()` - this type of action gets `distinct pairs` from the batch of users `User1` and `User2` which are tuples `{Pid, Data}` with `Pid`s passed by users calling `amoc_coordinator:add/2` or `amoc_coordinator:add/3`;
+
+`Coordination Actions` are executed in a reversed order in relation to their declaration in the `Coordination Plan`.
```
Comment (DenysGonchar) [inline doc/amoc_coordinator.md]: this is not true. the order of actions in one item is guaranteed (direct order, not reversed). the order of actions from different coordination items is not guaranteed at all (casted and executed in different processes). timeout/reset/stop events are not casted but ‘called’, so for the timeout/reset/stop evenst it’s guaranteed that all the coordination items with `all` are executed after all numbered coordination items are done.

File: doc/amoc_coordinator.md [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+% This action triggers sending {hello,Id} to the users 1 and 2
+Two new users showed up: Event = coordinate; ListOfUsers = [{<0.1142.0>,2},{<0.1140.0>,1}]
+Two new users showed up: Event = coordinate
+
+% Users 1 and 2 received messages and print them:
+{Msg = {hello,2}, Id = 2
+{Msg = {hello,1}, Id = 1
+
+User = 3
+% We have 3 users added to amoc_coordinator so all of actions {3, _} are triggered:
```
Comment (NelsonVides) [inline doc/amoc_coordinator.md]: > so all of the `{3, _}` actions [...]

(ordering of the english genitive, and one article)
(same in the sentence a few lines below)

File: doc/amoc_coordinator.md [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+Three new users showed up
+
+User = 4
+% We have 4 and 4 rem 2 == 0 therefore users added to amoc_coordinator so all of actions {2, _} are triggered:
```
Comment (NelsonVides) [inline doc/amoc_coordinator.md]: Same as above

**Comment by erszcz** [comment]
In general, the doc looks great!

I have one remark with regard to styling the text in documentation, both in this document, and in some other documentation that we write.

I like to follow the documentation style where `inline code` is similar to

```
code blocks
```

because it suggests "copy and paste": to a terminal or to a file with source code.

On the other hand, there are _other ways_ to use **emphasis in text**, and Good Reasons to use them, for example when naming things and suggesting: _That's what we say when we speak about new concepts._

TL;DR:

- Command / code intended for copy pasting? Use backticks: "echo hello world"
- Putting emphasis / introducing concept names / expanding acronyms? Use other means

This makes it easy to skim over when following a tutorial and pick out the commands from the commentary.

File: doc/amoc_coordinator.md [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+Two new users showed up: Event = coordinate; User1 = {<0.1144.0>,4}; User2 = {<0.1143.0>,3}
+Two new users showed up: Event = coordinate; ListOfUsers = [{<0.1144.0>,4},{<0.1143.0>,3}]
+Two new users showed up: Event = coordinate
+
+{Msg = {hello,4}, Id = 4
+{Msg = {hello,3}, Id = 3
+User = 5
+
+% You need to wait for a while, and ...
+% Timeout has been reached. Which triggers all of the Coordination Actions with remaining number of users remaining.
```
Comment (NelsonVides) [inline doc/amoc_coordinator.md]: "remaining" twice? 😛 

>  % Timeout has been reached, wiich triggers all of the Coordination Actions with the remaining number of users.

File: doc/amoc_coordinator.md [hunk 15]
```diff
@@ -0,0 +1,170 @@
...
+ - `fun(Event) -> any()` - this type of action does not care about particular users, but only about the number of them;
+ - `fun(Event, ListOfUsersData) -> any()` - this type of action gets `ListOfUsersData` which is a list of tuples `{Pid, Data}` with `Pid`s passed by users calling `amoc_coordinator:add/2` or `amoc_coordinator:add/3`;
+- `fun(Event, User1, User2) -> any()` - this type of action gets `distinct pairs` from the batch of users `User1` and `User2` which are tuples `{Pid, Data}` with `Pid`s passed by users calling `amoc_coordinator:add/2` or `amoc_coordinator:add/3`;
+
+`Coordination Actions` are executed in a reversed order in relation to their declaration in the `Coordination Plan`.
+
+`distinct pairs` - in the context, these are pairs from collection where:
+ - when `{A, B}` is in the `distinct pairs` then `{B, A}` is not;
+ - when `{A, A}` is not in the `distinct pairs`;
+ - all pairs are distinct;
```
[the remark this answers is not in the document]
Comment (aleklisi) [reply doc/amoc_coordinator.md]: ??

File: doc/amoc_coordinator.md (hunk 8 above)
Comment (erszcz) [reply doc/amoc_coordinator.md]: @NelsonVides this is a type definition, so the users (the instances of data) might actually be different, but the type is the same nonetheless.

File: doc/amoc_coordinator.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (NelsonVides) [reply doc/amoc_coordinator.md]: I just found it quite redundant, but I actually have no better suggestion 🤷‍♂

File: doc/amoc_coordinator.md (hunk 8 above)
Comment (erszcz) [reply doc/amoc_coordinator.md]:
```
$ erl -pa _build/default/lib/amoc/ebin/
Erlang/OTP 21 [erts-10.3.5.6] [source] [64-bit] [smp:4:4] [ds:4:4:10] [async-threads:1]

Enabled docsh 0.7.1 from: /Users/erszcz/work/erszcz/docsh/_build/default/lib/docsh
Call h(docsh) for interactive help.

Eshell V10.3.5.6  (abort with ^G)
1> l(amoc_coordinator).
{module,amoc_coordinator}
2> h(amoc_coordinator).
Module description is not available.
```

:'(

```
ok
3> t(amoc_coordinator).
-type coordination_action() :: fun((coordination_event(),
                                    [coordination_data()]) -> any()) |
                               fun((coordination_event(),
                                    maybe_coordination_data(),
                                    maybe_coordination_data()) -> any()) |
                               fun((coordination_event()) -> any()).
-type coordination_actions() :: [coordination_action()] | coordination_action().
-type coordination_data() :: {pid(), Data :: any()}.
-type coordination_event() :: coordinate | timeout | stop | reset.
-type coordination_item() :: {NoOfUsers :: pos_integer() | all,
                              coordination_actions()}.
-type coordination_plan() :: [coordination_item()] | coordination_item().
-type coordination_timeout_in_sec() :: pos_integer() | infinity.
-type maybe_coordination_data() :: coordination_data() | undefined.
-type normalized_coordination_item() :: {NoOfUsers :: pos_integer() | all,
                                         [coordination_action()]}.
-type state() :: {worker, pid()} | {timeout, pid()}.

ok
4>
```

:+1:
